### PR TITLE
cmd/puppeth: integrate blockscout

### PR DESCRIPTION
**Pull request by rjl493456442**
This PR integrate an open-source blockchain explorer `blockscout` into the puppeth.

Now puppeth can provide explorer service for PoA based chain and PoW based chain

### CODE REVIEW

**Comment by 5chdn** [comment]
That's pretty cool

**Comment by rocammo** [comment]
It works pretty well and it has a very good interface. Good job!

**Comment by hadv** [comment]
@rjl493456442 I want to build my own docker but facing this error. Do you have any idea? Thank you!

```
* Getting absinthe_phoenix (git@github.com:ayrat555/absinthe_phoenix.git)
fatal: cannot run ssh: No such file or directory
fatal: unable to fork
** (Mix) Command "git --git-dir=.git fetch --force --quiet --progress" failed
```

**Comment by rjl493456442** [comment]
@hadv It is an upstream issue. I've already opened an issue here https://github.com/poanetwork/blockscout/issues/1478

**Comment by hadv** [comment]
> @hadv It is an upstream issue. I've already opened an issue here [poanetwork/blockscout#1478](https://github.com/poanetwork/blockscout/issues/1478)

thanks, so the work-around is to replace it with https? And how can use build the puppeth/blockscout:latest image?

**Comment by hadv** [comment]
> It works pretty well and it has a very good interface. Good job!

yeah, great! Except I cannot run the data node with `--ws` options. the data node only can sync the data if I disable the **web-socket**. Any idea, @rjl493456442 ? Thanks!

**Comment by rocammo** [comment]
> > It works pretty well and it has a very good interface. Good job!
> 
> yeah, great! Except I cannot run the data node with `--ws` options. the data node only can sync the data if I disable the **web-socket**. Otherwise, the data node can only sync some blocks then drop all peers. Any idea, @rjl493456442 ? Thanks!

I don't know if there is any difference in whether the network is PoA or PoW. In my case, my implementation has been carried out in a network configured as PoW for a demo, and the truth is that I have not had any problems.

**Comment by hadv** [comment]
> I don't know if there is any difference in whether the network is PoA or PoW. In my case, my implementation has been carried out in a network configured as PoW for a demo, and the truth is that I have not had any problems.

I'm running a private PoA network with some divert then I need to build my own blockscout docker image and I have to remove compile's `--warnings-as-errors` flag to build the blockscout project.

**Comment by rocammo** [comment]
> > I don't know if there is any difference in whether the network is PoA or PoW. In my case, my implementation has been carried out in a network configured as PoW for a demo, and the truth is that I have not had any problems.
> 
> I'm running a private PoA network with some divert then I need to build my own blockscout docker image and I have to remove compile's `--warnings-as-errors` flag to build the blockscout project also by the way.

I didn't have any compilation problems when I added it just two days ago and I also compiled it from source, so it's strange that you're having it. Could you please provide the explorer container log by `docker logs <containerId>` or the compilation errors that you're having?

**Comment by hadv** [comment]
> I didn't have any compilation problems when I added it just two days ago and I also compiled it from source, so it's strange that you're having it. Could you please provide the explorer container log by `docker logs <containerId>` or the compilation errors that you're having?

I cannot build the docker image with warning as error option

https://github.com/puppeth/blockscout/blob/9af22a89de29c094acc2349cd5b368a5564fe2ee/apps/block_scout_web/mix.exs#L142

**Comment by hadv** [comment]
@rocammo can you try to run with contract metadata and validator? I enable them but the explorer shutdown out of the blue 😢 

```
if System.get_env("METADATA_CONTRACT") && System.get_env("VALIDATORS_CONTRACT") do
  config :explorer, Explorer.Validator.MetadataRetriever,
    metadata_contract_address: System.get_env("METADATA_CONTRACT"),
    validators_contract_address: System.get_env("VALIDATORS_CONTRACT")

  config :explorer, Explorer.Validator.MetadataProcessor, enabled: true
else
  config :explorer, Explorer.Validator.MetadataProcessor, enabled: false
end
```

**Comment by rjl493456442** [comment]
Btw there is a known issue in the current PR found by @mirooy and @Kimjingoo that now the geth node version in blockscout image is different with the `wallet` image or some other components. So that there could have a consensus error now if you try to deploy a smart contract.

Have to wait the upstream fix the build issue and then rebuild the `puppeth/blockscout`.

**Comment by hadv** [comment]
> Btw there is a known issue in the current PR found by @mirooy and @Kimjingoo that now the geth node version in blockscout image is different with the `wallet` image or some other components. So that there could have a consensus error now if you try to deploy a smart contract.
> 
> Have to wait the upstream to fix the build issue and then rebuild the `puppeth/blockscout`.

is this the error?

```

** (FunctionClauseError) no function clause matching in Logger.level_to_number/1
    (logger) lib/logger.ex:536: Logger.level_to_number(:prod)
    (logger) lib/logger.ex:533: Logger.compare_levels/2
    (logger_file_backend) lib/logger_file_backend.ex:33: LoggerFileBackend.handle_event/2
    (stdlib) gen_event.erl:577: :gen_event.server_update/4
    (stdlib) gen_event.erl:559: :gen_event.server_notify/4
    (stdlib) gen_event.erl:300: :gen_event.handle_msg/6
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Last message: {:debug, #PID<0.499.0>, {Logger, ["QUERY", 32, "OK", "", [32, "db", 61, '1.5', 109, 115], [], [], 10, "commit", 32, "[]"], {{2019, 3, 2}, {3, 11, 47, 758}}, [pid: #PID<0.1086.0>, ansi_color: :magenta, line: 50, function: "log/3", module: Ecto.LogEntry, file: "lib/ecto/log_entry.ex", application: :ecto, transaction_id: 1991, import_id: 1959, block_number: 135, fetcher: :block_realtime]}}
State: %{format: [:time, " ", :metadata, "[", :level, "] ", :message, "\n"], inode: nil, io_device: nil, level: :prod, metadata: [], metadata_filter: [fetcher: :failed_created_addresses], name: :failed_contract_creations, path: "/opt/app/logs/prod/indexer/failed_contract_creations.log", rotate: nil}
2019-03-02T03:11:47.786 application=indexer [info] refetch from consensus was found on block (60). A reorg initiated.
2019-03-02T03:11:47.800 [error] GenServer #PID<0.1186.0> terminating
** (stop) {:EXIT, {:function_clause, [{Logger, :level_to_number, [:prod], [file: 'lib/logger.ex', line: 536]}, {Logger, :compare_levels, 2, [file: 'lib/logger.ex', line: 533]}, {LoggerFileBackend, :handle_event, 2, [file: 'lib/logger_file_backend.ex', line: 33]}, {:gen_event, :server_update, 4, [file: 'gen_event.erl', line: 577]}, {:gen_event, :server_notify, 4, [file: 'gen_event.erl', line: 559]}, {:gen_event, :handle_msg, 6, [file: 'gen_event.erl', line: 300]}, {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 249]}]}}
Last message: {:gen_event_EXIT, {LoggerFileBackend, :failed_contract_creations}, {:EXIT, {:function_clause, [{Logger, :level_to_number, [:prod], [file: 'lib/logger.ex', line: 536]}, {Logger, :compare_levels, 2, [file: 'lib/logger.ex', line: 533]}, {LoggerFileBackend, :handle_event, 2, [file: 'lib/logger_file_backend.ex', line: 33]}, {:gen_event, :server_update, 4, [file: 'gen_event.erl', line: 577]}, {:gen_event, :server_notify, 4, [file: 'gen_event.erl', line: 559]}, {:gen_event, :handle_msg, 6, [file: 'gen_event.erl', line: 300]}, {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 249]}]}}}
State: {Logger, {LoggerFileBackend, :failed_contract_creations}}
```

**Comment by hadv** [comment]
> > > I don't know if there is any difference in whether the network is PoA or PoW. In my case, my implementation has been carried out in a network configured as PoW for a demo, and the truth is that I have not had any problems.
> > 
> > 
> > I'm running a private PoA network with some divert then I need to build my own blockscout docker image and I have to remove compile's `--warnings-as-errors` flag to build the blockscout project also by the way.
> 
> I didn't have any compilation problems when I added it just two days ago and I also compiled it from source, so it's strange that you're having it. Could you please provide the explorer container log by `docker logs <containerId>` or the compilation errors that you're having?

This is exactly the problem I have and they provide a work around for it

https://github.com/poanetwork/blockscout/issues/1391

**Comment by hadv** [comment]
> > It works pretty well and it has a very good interface. Good job!
> 
> yeah, great! Except I cannot run the data node with `--ws` options. the data node only can sync the data if I disable the **web-socket**. Otherwise, the data node can only sync some blocks then drop all peers. Any idea, @rjl493456442 ? Thanks!

the docker container shutdown suddenly if connect to local rpc/ws endpoint. If I connect to other endpoint it's okay. Seems problem due to running every things on the same docker container

File: cmd/puppeth/module_explorer.go [hunk 1]
```diff
@@ -30,107 +30,80 @@ import (
...
-ADD chain.json /chain.json
+FROM puppeth/blockscout:latest
 
+ADD genesis.json /genesis.json
 RUN \
-  echo '(cd ../eth-net-intelligence-api && pm2 start /ethstats.json)' >  explorer.sh && \
-	echo '(cd ../etherchain-light && npm start &)'                      >> explorer.sh && \
-	echo 'exec /parity/parity --chain=/chain.json --port={{.NodePort}} --tracing=on --fat-db=on --pruning=archive' >> explorer.sh
+  echo 'geth --cache 512 init /genesis.json' > explorer.sh && \
+	echo $'exec geth --networkid {{.NetworkID}} --syncmode "full"  --port {{.NodePort}} --bootnodes {{.Bootnodes}} --ethstats \'{{.Ethstats}}\' --cache=512 --rpc --rpccorsdomain "*" --rpcvhosts "*" --ws --wsorigins "*" &' >> explorer.sh && \
```
Comment (hadv) [inline cmd/puppeth/module_explorer.go]: we should enable `debug` rpc api to reading smart contract & internal transaction

Comment (hadv) [inline cmd/puppeth/module_explorer.go]: I prefer to allow user to specify which geth node they want to connect to other than always using localhost

Comment (rjl493456442) [reply cmd/puppeth/module_explorer.go]: Yes, I added it in my local and haven't pushed it. Thanks for catching!

Comment (rjl493456442) [reply cmd/puppeth/module_explorer.go]: My original idea is: this embedded geth node is only available for blockscout. Probably if a user wants to connect to a geth node, then they can deploy a new node somehow.

Comment (hadv) [reply cmd/puppeth/module_explorer.go]: Yes, it's a good idea. by the way, when I running embedded geth node the docker container suddenly shutdown. I might be wrong somewhere but still dont know the reason but when connect to outside node, it worked properly without any problem.

File: cmd/puppeth/wizard_explorer.go [hunk 2]
```diff
@@ -75,12 +67,21 @@ func (w *wizard) deployExplorer() {
...
 		infos.datadir = w.readDefaultString(infos.datadir)
 	}
+	// Figure out where the user wants to store the persistent data for backend database
+	fmt.Println()
+	if infos.dbdir == "" {
+		fmt.Printf("Where should postgres data be stored on the remote machine?\n")
+		infos.dbdir = w.readString()
+	} else {
+		fmt.Printf("Where should postgres data be stored on the remote machine? (default = %s)\n", infos.dbdir)
+		infos.dbdir = w.readDefaultString(infos.datadir)
```
Comment (hadv) [inline cmd/puppeth/wizard_explorer.go]: typo seems 

```suggestion
		infos.dbdir = w.readDefaultString(infos.dbdir)
```

**Comment by rjl493456442** [comment]
Fix the blockscout docker image and now it can work in the `clique` mode. Yeah


image

**Comment by rjl493456442** [comment]
Now the upstream project is much stabler. Appreciate for the support from @ayrat555 and I think this PR is review ready :))

**Comment by kant777** [comment]
How to run blockscout after this integration?

**Pull request merged**